isolation_test.go

```diff
@@ -0,0 +1,250 @@
+	session1.MustExec("commit;")
+	session2.MustExec("commit;")
+}
+
+func (s *testIsolationSuite) TestP2NonRepeatableRead(c *C) {
+	session1 := testkit.NewTestKitWithInit(c, s.store)
+	session2 := testkit.NewTestKitWithInit(c, s.store)
+
+	session1.MustExec("drop table if exists x;")
+	session1.MustExec("drop table if exists y;")
+	session1.MustExec("create table x (id int primary key, c int);")
+	session1.MustExec("insert into x values(1, 1);")
+	session1.MustExec("create table y (id int primary key, c int);")
+	session1.MustExec("insert into y values(1, 1);")
+
+	session1.MustExec("begin;")
+	session2.MustExec("begin;")
+	session1.MustQuery("select c from x where id = 1;").Check(testkit.Rows("1"))
+	session2.MustQuery("select c from x where id = 1;").Check(testkit.Rows("1"))
+	session2.MustExec("update x set c = c+1 where id = 1;")
+	session2.MustQuery("select c from y where id = 1;").Check(testkit.Rows("1"))
+	session2.MustExec("update y set c = c+1 where id = 1;")
+	session2.MustExec("commit;")
+	session1.MustQuery("select c from y where id = 1;").Check(testkit.Rows("1"))
+	session1.MustExec("commit;")
+}
+
+func (s *testIsolationSuite) TestP3Phantom(c *C) {
+	session1 := testkit.NewTestKitWithInit(c, s.store)
+	session2 := testkit.NewTestKitWithInit(c, s.store)
+
+	session1.MustExec("drop table if exists x;")
+	session1.MustExec("drop table if exists z;")
+	session1.MustExec("create table x (id int primary key, c int);")
+	session1.MustExec("insert into x values(1, 1);")
+	session1.MustExec("create table z (id int primary key, c int);")
+	session1.MustExec("insert into z values(1, 1);")
+
+	session1.MustExec("begin;")
+	session2.MustExec("begin;")
+	session1.MustQuery("select c from x where id < 5;").Check(testkit.Rows("1"))
+	session2.MustExec("insert into x values(2, 1);")
+	session2.MustQuery("select c from z where id = 1;").Check(testkit.Rows("1"))
+	session2.MustExec("update z set c = c+1 where id = 1;")
+	session2.MustExec("commit;")
+	session1.MustQuery("select c from z where id = 1;").Check(testkit.Rows("1"))
+	session1.MustExec("commit;")
+}
+
+func (s *testIsolationSuite) TestP4LostUpdate(c *C) {
+	session1 := testkit.NewTestKitWithInit(c, s.store)
+	session2 := testkit.NewTestKitWithInit(c, s.store)
+
+	session1.MustExec("drop table if exists x;")
+	session1.MustExec("create table x (id int primary key, c int);")
+	session1.MustExec("insert into x values(1, 1);")
+
+	session1.MustExec("begin;")
+	session1.MustQuery("select c from x where id = 1;").Check(testkit.Rows("1"))
+	session2.MustExec("begin;")
+	session2.MustQuery("select c from x where id = 1;").Check(testkit.Rows("1"))
+	session2.MustExec("update x set c = c+1 where id = 1;")
+	session2.MustExec("commit;")
+	session1.MustExec("update x set c = c+1 where id = 1;")
+	_, err := session1.Exec("commit;")
+	c.Assert(err, NotNil)
+}
+
+// cursor is not supported
+func (s *testIsolationSuite) TestP4CLostUpdate(c *C) {}
```

Comment:
??

Reply:
P4C is used for cursor which tidb doesn't support